File: C++/105.cpp
#include <vector>
#include <cstdlib>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/*
思路：先从先序遍历中确定根节点，再从中序遍历中确定左右树的长度，其中先序遍历中根节点的下一个
便是左节点，再经过左子树长度的节点之后，便是右节点；
依次类推
*/
class Solution {
public:
    TreeNode *buildCore(vector<int> &preorder, int preStart, int preEnd, vector<int> &inorder, int inStart, int inEnd) {
        //前序遍历第一个节点是根节点
        int rootValue = preorder.at(preStart);
        TreeNode *root = new TreeNode(rootValue);

        //前序序列只有根节点
        if (preStart == preEnd) {
            return root;
        }
            
        //遍历中序序列，找到根节点的位置
        int rootInorder = inStart;
        while (inorder.at(rootInorder)!=rootValue && rootInorder<=inEnd) {
            rootInorder++;
        }

        //左子树的长度
        int leftLength = rootInorder-inStart;
        //前序序列中左子树的最后一个节点
        int leftPreEnd = preStart+leftLength;
        int leftInEnd = inStart+leftLength;

        //左子树非空
        if (leftLength > 0) {
            //重建左子树
            root->left = buildCore(preorder, preStart+1, leftPreEnd, inorder, inStart, leftInEnd);
        }
        //右子树非空
        if (leftLength < preEnd-preStart) {
            //重建右子树
            root->right = buildCore(preorder, leftPreEnd+1, preEnd, inorder, rootInorder+1, inEnd);
        }
        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.size() == 0)
            return NULL;
        return buildCore(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1);
    }
};

File: Python/105.py
from typing import List
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    
    def buildCore(self, preorder: List[int], preStart: int, preEnd: int, inorder: List[int], inStart: int, inEnd: int) -> TreeNode:
        # 根节点
        root = TreeNode(preorder[preStart])

        # 是否长度为1
        if preStart==preEnd:
            return root
        
        # 中序遍历中得到左子树的长度
        rootIndex = inStart
        while preorder[preStart]!=inorder[rootIndex] and rootIndex<inEnd:
            rootIndex += 1
        leftLength = rootIndex-inStart

        #重建左右子树
        if leftLength > 0:
            root.left = self.buildCore(preorder, preStart+1, preStart+leftLength, inorder, inStart, rootIndex-1)
        if leftLength < preEnd-preStart:
            root.right = self.buildCore(preorder, preStart+leftLength+1, preEnd, inorder, rootIndex+1, inEnd)
        
        return root

    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if preorder is None or inorder is None or len(preorder) == 0:
            return None
        return self.buildCore(preorder, 0, len(preorder)-1, inorder, 0, len(inorder)-1)
